File: 53.maximum-subarray.cpp
/*
 * @lc app=leetcode id=53 lang=cpp
 *
 * [53] Maximum Subarray
 */

// @lc code=start
#include<vector>
#include<limits.h>
using namespace std;
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        vector<int> max_sums(nums.size()+1,0);
        int max_sum = INT_MIN;
        for(int i = nums.size()-1; i >= 0; i--){
            //[i] -> [i+1] + cur
            max_sums[i] = ( (max_sums[i+1] > 0)?max_sums[i+1]:0 ) + nums[i];
            if(max_sums[i] > max_sum)
                max_sum = max_sums[i];
        }

        return max_sum;
    }
};
// @lc code=end

File: 53.maximum-subarray.go
/*
 * @lc app=leetcode id=53 lang=golang
 *
 * [53] Maximum Subarray
 */

// @lc code=start
var memo map[int]int

var _nums []int

func maxSubArray(nums []int) int {
	_nums = nums
	memo = make(map[int]int, 0)

	msum := math.MinInt
	for i := 0; i < len(_nums); i++ {
		msum = max(msum, dp(i))
	}

	return msum
}

func dp(i int) (msum int) {
	if i < 0 {
		return 0
	} else if i == 0 {
		return _nums[i]
	}

	if v, ok := memo[i]; ok {
		return v
	}

	msum = max(dp(i-1)+_nums[i], _nums[i])

	memo[i] = msum
	return
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

// @lc code=end
